records/2024-05-23.scalability.short-code-implementation.md

```diff
@@ -0,0 +1,107 @@
+1. A template can be of type PRIORITY, NORMAL, and BULK which ***currently*** maps directly to the HIGH, MEDIUM, and LOW queues. 
+1. The templates refer to both SMS and Email templates, but for shortcodes, this ADR focuses on SMS use cases
+
+# Implementation
+
+## Template Categorization
+
+When a client creates a template we want to be able to Categorize it into a certain type of template. We will start with the following Template Categories that were created through research on Notify's current templates.
+
+| Template Category | Example  |
+| ----------------- | -------- |
+| Status Update | Notice of change in status<br>The progress of a submission |
+| <br>Promotional call to action | Surveys<br>General apply now<br>"Learn more" |
+| Service-related requests | Submit additional documents<br>Follow up to move a process forward |
+| Fulfillment with attachments - email only | "Here’s your permit" |
+| <br>Broadcast messages | General information - (not related to transactions such as COVID-19 information)<br>Newsletters<br>Status updates |
+| <br>Auto-reply | No-reply messages<br>Acknowledgments<br>Response wait times |
+| Verification message | Authentication codes <br>Confirmation of account changes |
+| Confirmation / Receipts | Record of transactions<br>Approvals |
+| Other | User defined |
+
+### Determining Template Priority
+A new table will be added to the DB for TemplateCategories. The primary key from TemplateCategories will be a foreign key to the Templates Table. All TemplateCategories will be assigned a pre-computed priority.
+
+<!-- ![Potential ERD](https://raw.githubusercontent.com/cds-snc/notification-adr/records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD") -->
+
+![Potential ERD](../records/diagrams/2024-05-23.scalability.short-code-implementation/ERD.jpg?raw=true "Potential ERD")
+
+```
+process_type is (bulk, normal, priority)
+```
+
+Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides, an Admin will edit the process_type on the Templates Table. For this to work, we have two options:
+1. Create a new column in Templates `new_process_type`
+2. Clear the current process_type in the Templates table to be empty.
+
+After the above is implemented,
+```
+Notification Priority = Template.new_process_type else TemplateCategory.process_type
+```
+For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.
+
+**Use case:**  
+An SMS Template is of type "XYZ" that is of "low" priority. The client reaches out to us about the importance of the messages being sent immediately. We leave the category of the Template as "XYZ", but manually update the Template to "high" priority. This means that the message will be sent using the short-code.
```

Comment:
Most likely the short code, but there are instances where that might not be the case, i.e. if the destination code is American based, or the service has an associated long code -- or several even, the pools will allow us to do grouping that we could not before.